File: contests/w22/SequentialPrefixFunction.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
/// <summary>
/// https://www.hackerrank.com/contests/w22/challenges/sequential-prefix-function
/// </summary>
class Solution6
{
    static void Main(String[] args)
    {
        TextReader tIn = Console.In;
        TextWriter tOut = Console.Out;

        int Q = int.Parse(tIn.ReadLine());

        int[] S = new int[Q];
        int ix = 0;

        int[] LPS = new int[Q];
        int[] LEN = new int[Q];
        int len = 0;

        for (int q = 0; q < Q; q++)
        {
            string line = tIn.ReadLine();
            switch (line[0])
            {
                case '+':
                    S[ix] = int.Parse(line.Substring(2));

                    if (ix > 0)
                        while (true)
                        {
                            if (S[ix] == S[len])
                            {
                                len++;
                                LPS[ix] = len;
                                break;
                            }
                            else
                            {
                                if (len != 0)
                                {
                                    len = LPS[len - 1];
                                }
                                else
                                {
                                    LPS[ix] = 0;
                                    break;
                                }
                            }
                        }
                    LEN[ix] = len;
                    tOut.WriteLine(LPS[ix]);
                    ix++;
                    break;
                case '-':
                    ix--;
                    len = ix > 0 ? LEN[ix - 1] : 0;
                    tOut.WriteLine(len);
                    break;
            }
        }
    }
}

File: contests/w22/SequentialPrefixFunction.vb
Imports System
Imports System.IO
Module Module1

    Sub Main()

        Dim tIn As TextReader = Console.In
        Dim tOut As TextWriter = Console.Out

        Dim nQ As Integer = Integer.Parse(tIn.ReadLine())

        Dim S(nQ + 1) As Integer
        Dim ix As Integer = 0

        Dim LPS(nQ + 1) As Integer
        Dim LEN(nQ + 1) As Integer
        Dim alen As Integer = 0

        For q As Integer = 0 To nQ - 1
            Dim line As String = tIn.ReadLine()
            Select Case line(0)
                Case "+"c
                    S(ix) = Integer.Parse(line.Substring(2))

                    If ix > 0 Then
                        While True
                            If (S(ix) = S(alen)) Then
                                alen += 1
                                LPS(ix) = alen
                                Exit While
                            Else
                                If alen <> 0 Then
                                    alen = LPS(alen - 1)
                                Else
                                    LPS(ix) = 0
                                    Exit While
                                End If
                            End If
                        End While
                    End If
                    LEN(ix) = alen
                    tOut.WriteLine(LPS(ix))
                    ix += 1
                Case "-"c
                    ix -= 1
                    alen = If(ix > 0, LEN(ix - 1), 0)
                    tOut.WriteLine(alen)
            End Select
        Next
    End Sub

End Module
